controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,418 @@
+	if err != nil {
+		if errors.Is(err, ErrNotFoundDiscovery) {
+			return reconcile.Result{RequeueAfter: time.Minute * 3}, nil
+		}
+
+		log.Error(err, "Error to get matching kind and apigroup")
+
+		return reconcile.Result{}, nil
+	}
+
+	r.addConstraintToMap(constraint.GetName(), *constraintSyncOnlyEntries)
+	updatedSyncOnly := r.getUniqSyncOnly()
+	err = r.updateConfigSyncOnlyEntry(ctx, config, updatedSyncOnly)
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ConstraintStatusReconciler) getSyncOnlys(constraintMatchKinds []interface{}) (
+	*[]v1alpha1.SyncOnlyEntry, error,
+) {
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+
+	for _, match := range constraintMatchKinds {
+		newKind, ok := match.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		apiGroups, ok := newKind["apiGroups"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		kindsInKinds, ok := newKind["kinds"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		for _, apiGroup := range apiGroups {
+			for _, kind := range kindsInKinds {
+				version, err := r.getAPIVersion(kind.(string), apiGroup.(string), false)
+				if err != nil {
+					return nil, err
+				}
+
+				syncOnlys = append(syncOnlys, v1alpha1.SyncOnlyEntry{
+					Group:   apiGroup.(string),
+					Version: version,
+					Kind:    kind.(string),
+				})
+			}
+		}
+	}
+
+	return &syncOnlys, nil
+}
+
+func (r *ConstraintStatusReconciler) getAPIVersion(kind string, apiGroup string, skipRefresh bool) (string, error) {
+	// cool time(10 min) to refresh discoveries
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		r.discoveryLastRefreshed = time.Now()
+
+		if err != nil {
+			return "", err
+		}
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison, err := schema.ParseGroupVersion(resc.GroupVersion)
+		if err != nil {
+			r.Log.Error(err, "Cannot Parse Group and version in getApiVersion ", "GroupVersion:", resc.GroupVersion)
+
+			continue
+		}
+
+		group := groupVerison.Group
+		version := groupVerison.Version
+		// Consider groupversion == v1 or groupversion == app1/v1
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version, nil
+			}
+		}
+	}
+
+	if !skipRefresh {
+		// Get new discoveryInfo, when any resource is not found
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		r.discoveryLastRefreshed = time.Now()
+		// Retry one more time after refresh the discovery
+		return r.getAPIVersion(kind, apiGroup, true)
+	}
+
+	return "", ErrNotFoundDiscovery
+}
+
+// Retrieve all groups and versions to add in config sync
+// Constraints present only kind and group so this function helps to find the version
+func (r *ConstraintStatusReconciler) refreshDiscoveryInfo() error {
+	discoveryClient := r.ClientSet.Discovery()
```

Comment:
Could you please use a RWMutex for refreshing and accessing the discovery information? Otherwise you could have several discovery refresh requests occur all at one and you could get inconsistent results if `r.apiResourceList` changes when it's being used by another goroutine.

Reply:
since I change to MaxConcurrentReconciles: int(1), we don't need to change right?